lua/extn/sky/unstable/io/arc.lua

```diff
@@ -46,24 +52,109 @@ function ArcDialGesture:new(props)
 end
 
 function ArcDialGesture:process(event, output, state)
-  if sky.is_type(event, ArcInput.ARC_ENC_EVENT) then
-    if event.n == self.which then
-      local range = self.max - self.min
-      local inc = range / self.steps
-      local change = inc * event.delta * self.scale
-      local next = util.clamp(self._value + change, self.min, self.max)
-      if next ~= self._value then
-        self._value = next
-        output({
-          type = self.ARC_DIAL_EVENT,
-          arc = event.arc,
-          n = event.n,
-          value = self._value,
-          normalized = self._value / range,
-        })
+  if sky.is_type(event, ArcInput.ARC_ENC_EVENT) and event.n == self.which then
+    local range = self.max - self.min
+    local inc = range / self.steps
+    local change = inc * event.delta * self.scale
+    local next = util.clamp(self._value + change, self.min, self.max)
+    if next ~= self._value then
+      self._value = next
+      output(self.mk_dial(event.n, self._value, self._value / range, event.arc))
+    end
+  else
+    -- if not an arc enc event pass on through
+    output(event)
+  end
+end
+
+function ArcDialGesture.mk_dial(n, value, normalized, arc)
+  return {
+    type = ArcDialGesture.ARC_DIAL_EVENT,
+    arc = arc,
+    n = n,
+    value = value,
+    normalized = normalized,
+  }
+end
+
+function ArcDialGesture.is_dial(event)
+  return sky.is_type(event, ArcDialGesture.ARC_DIAL_EVENT)
+end
+
+--
+-- ArcDialSmoother
+--
+
+local ArcDialSmoother = sky.Device:extend()
+ArcDialSmoother.STEP_EVENT = 'ARC_DIAL_SMOOTHER_STEP'
+
+function ArcDialSmoother:new(props)
+  ArcDialSmoother.super.new(self, props)
+  self.which = props.which or 1
+  self._time = props.time or 1
+  self._sr = props.sr or 5
+  self._scheduler = nil
+  self._smoother = filters.smoother.new(self._time, self._sr)
+  self._smoothing = false
+  self._target = nil
+
+  self._step = { type = self.STEP_EVENT }
+end
+
+function ArcDialSmoother:device_inserted(chain)
+  if self._scheduler ~= nil then
+    error('ArcDialSmoother: one instance cannot be used in multiple chains at the same time')
+  end
+  self._scheduler = chain:scheduler(self)
+end
+
+function ArcDialSmoother:device_removed(chain)
+  self._scheduler = nil
+end
+
+function ArcDialSmoother:set_time(t)
+  self._time = t
+  self._smoother:set_time(t)
+end
+
+function ArcDialSmoother:set_sr(sr)
+  self._sr = sr
+  -- FIXME: need to implement this
```

Comment:
What's missing from this? Updating the smoother?

Reply:
ack, meant to just delete that for now.